### PR TITLE
DAOS-7353 swim: rewrite to using regular two-ways RPC with reply

### PR DESCRIPTION
**Pull request by dmiter**
Completely rework SWIM code to use regular two-ways RPC with reply.

### CODE REVIEW

File: src/cart/crt_swim.c [hunk 1]
```diff
@@ -191,46 +156,151 @@ static void crt_swim_srv_cb(crt_rpc_t *rpc_req)
...
+				D_TRACE_DEBUG(DB_TRACE, rpc,
+					      "suspend %s with x updates. "
+					      "%lu: %lu <= %lu\n",
+					      SWIM_RPC_TYPE_STR[rpc_type],
+					      self_id,
+					      (rpc_type == SWIM_RPC_PING)
+					      ? to_id : rpc_in->src,
+					      from_id);
+				/* Keep this RPC in ipings queue */
+				RPC_ADDREF(rpc_priv);
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c:280]: Yes, It's just sign we already have different IREQ in queue already. We always add it to queue. Therefore we need to increment the reference.

File: src/cart/crt_swim.c [hunk 2]
```diff
@@ -301,33 +446,77 @@ static int crt_swim_send_message(struct swim_context *ctx, swim_id_t to,
...
-	rc = crt_req_create(crt_ctx, &ep, opc, &rpc_req);
+	opc = CRT_PROTO_OPC(CRT_OPC_SWIM_BASE, CRT_OPC_SWIM_VERSION, rpc_type);
+	rc = crt_req_create(crt_ctx, &ep, opc, &rpc);
 	if (rc) {
 		D_ERROR("crt_req_create(): "DF_RC"\n", DP_RC(rc));
 		D_GOTO(out, rc);
 	}
 
-	rc = crt_req_set_timeout(rpc_req, crt_swim_rpc_timeout);
+	timeout = 1 + swim_ping_timeout_get() / NSEC_PER_USEC;
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c:456]: Yes, exactly. Ok, I will add comment here.

File: src/cart/crt_swim.c [hunk 3]
```diff
@@ -301,33 +446,77 @@ static int crt_swim_send_message(struct swim_context *ctx, swim_id_t to,
...
 out:
-	if (rc && rpc_req != NULL)
-		crt_req_decref(rpc_req);
+	if (rc && rpc != NULL)
+		crt_req_decref(rpc);
+	return rc;
+}
+
+static int crt_swim_iping_expired(struct swim_context *ctx, swim_id_t from,
+				  swim_id_t to, void *args)
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c:485]: I need to think about this. It looks good but maybe same corner cases which will break this approach. Will think more...

File: src/cart/crt_swim.c [hunk 4]
```diff
@@ -301,33 +446,77 @@ static int crt_swim_send_message(struct swim_context *ctx, swim_id_t to,
...
 			      DP_RC(rc));
 		D_GOTO(out, rc);
 	}
 
-	rpc_swim_input = crt_req_get(rpc_req);
-	rpc_swim_input->src = self_id;
-	rpc_swim_input->upds.ca_arrays = upds;
-	rpc_swim_input->upds.ca_count  = nupds;
+	rpc_in = crt_req_get(rpc);
+	rpc_in->src = id;
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c:467]: rpc_in->src is used only for IREQ to know which target is suspected and should be indirectly pinged.

[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c]: done

File: src/cart/crt_swim.c [hunk 5]
```diff
@@ -33,7 +33,8 @@
 	((struct swim_member_update) (upds)		CRT_ARRAY)
 
 #define CRT_OSEQ_RPC_SWIM	/* output fields */		 \
-	((int32_t)		     (rc)		CRT_VAR)
+	((int32_t)		     (rc)		CRT_VAR) \
+	((struct swim_member_update) (upds)		CRT_ARRAY)
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c]: done

File: src/cart/crt_context.c [hunk 6]
```diff
@@ -473,6 +474,7 @@ crt_context_destroy(crt_context_t crt_ctx, int force)
 			D_GOTO(out, rc);
 	}
 
+	timeout = 1 + swim_ping_timeout_get() / NSEC_PER_USEC;
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_context.c]: done

File: src/cart/crt_swim.c [hunk 7]
```diff
@@ -191,46 +156,152 @@ static void crt_swim_srv_cb(crt_rpc_t *rpc_req)
...
+			break;
+		case SWIM_RPC_IREQ:
+			/*
+			 * Turn off RPC timeout because of it will be tracked
+			 * by ipings timeout through iping_expired() callback.
+			 */
+			rc = crt_req_set_timeout(rpc, UINT32_MAX);
+			if (rc)
+				D_TRACE_ERROR(rpc, "crt_req_set_timeout(): "
+					      DF_RC"\n", DP_RC(rc));
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.c]: removed.

File: src/cart/swim/swim.c [hunk 8]
```diff
@@ -286,16 +330,19 @@ static int
...
 
 	rc = ctx->sc_ops->get_member_state(ctx, id, &id_state);
 	if (rc) {
-		SWIM_ERROR("get_member_state(): "DF_RC"\n", DP_RC(rc));
+		SWIM_ERROR("get_member_state(%lu): "DF_RC"\n", id, DP_RC(rc));
 		D_GOTO(out, rc);
 	}
 
+	if (id_state.sms_status == SWIM_MEMBER_INACTIVE)
+		D_GOTO(out, rc = 0);
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/swim/swim.c:344]: This is done to avoid any state changes during INACTIVE state.

[the remark this answers is not in the document]
Comment (liw) [reply src/cart/swim/swim.c:344]: Hmm, how do we handle crashes of INACTIVE members? I thought we just defer the confirmation of INACTIVE deaths by a longer-than-usual timeout. And, if someone tells me an INACTIVE member has been confirmed as DEAD, I accept the news. That is no longer the design?

**Review by liw — APPROVED** [review]

**Review by liw — DISMISSED** [review]

**Comment by johannlombardi** [comment]
@frostedcmos please inspect.

**Review by liw — DISMISSED** [review]

**Comment by dmiter** [comment]
@frostedcmos could you inspect please?

File: src/cart/crt_context.c [hunk 9]
```diff
@@ -849,10 +851,34 @@ crt_context_timeout_check(struct crt_context *crt_ctx)
...
 	d_list_t			 timeout_list;
 	uint64_t			 ts_now;
-	int				 count = 0;
 
 	D_ASSERT(crt_ctx != NULL);
 
+	if (crt_initialized() && crt_is_service() && crt_gdata.cg_swim_inited) {
+		struct crt_grp_priv	*gp = crt_gdata.cg_grp->gg_primary_grp;
+		struct crt_swim_membs	*csm = &gp->gp_membs_swim;
+		swim_id_t		 self_id = swim_self_get(csm->csm_ctx);
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_context.c:879]: No, here we need only self_id and ctx index. They read only and don't care if they changed later.

File: src/cart/crt_context.c [hunk 10]
```diff
@@ -849,10 +851,34 @@ crt_context_timeout_check(struct crt_context *crt_ctx)
 	struct d_binheap_node		*bh_node;
 	d_list_t			 timeout_list;
 	uint64_t			 ts_now;
-	int				 count = 0;
 
 	D_ASSERT(crt_ctx != NULL);
 
+	if (crt_initialized() && crt_is_service() && crt_gdata.cg_swim_inited) {
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_context.c:876]: Agree, probably I should simplify this condition...

File: src/cart/crt_context.c [hunk 11]
```diff
@@ -849,10 +851,34 @@ crt_context_timeout_check(struct crt_context *crt_ctx)
...
+		struct crt_grp_priv	*gp = crt_gdata.cg_grp->gg_primary_grp;
+		struct crt_swim_membs	*csm = &gp->gp_membs_swim;
+		swim_id_t		 self_id = swim_self_get(csm->csm_ctx);
+
+		if (crt_ctx->cc_idx == csm->csm_crt_ctx_idx &&
+		    crt_ctx->cc_last_unpack_hlc != 0 &&
+		    self_id != SWIM_ID_INVALID) {
+			uint64_t delay = crt_hlc2msec(crt_hlc_get() -
+						   crt_ctx->cc_last_unpack_hlc);
+			uint64_t max_delay = swim_suspect_timeout_get() * 2 / 3;
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_context.c:886]: Ok. Will add comment.

File: src/cart/crt_swim.h [hunk 12]
```diff
@@ -29,13 +28,38 @@ struct crt_swim_membs {
 	D_CIRCLEQ_HEAD(, crt_swim_target) csm_head;
 	struct crt_swim_target		*csm_target;
 	struct swim_context		*csm_ctx;
-	uint64_t			 csm_hlc;
 	int				 csm_crt_ctx_idx;
 };
 
+static inline int
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/cart/crt_swim.h:35]: ok

**Review by liw — APPROVED** [review]

**Comment by dmiter** [comment]
@frostedcmos can we land this?

**Comment by dmiter** [comment]
@frostedcmos could you review this PR please?

**Review by frostedcmos — APPROVED** [review]
two minor comments; can be addressed in subsequent patch

File: src/include/cart/swim.h [hunk 13]
```diff
@@ -163,15 +178,43 @@ void swim_self_set(struct swim_context *ctx, swim_id_t self_id);
...
 			struct swim_member_update *upds, size_t nupds);
 
+/**
+ * Prepare a SWIM message for other group member.
+ *
+ * @param[in]  ctx	SWIM context pointer from swim_init()
+ * @param[in]  id	IDs of selected target for message
+ * @param[in]  to	IDs of message to send to
+ * @param[in]  upds	SWIM updates from other group member
+ * @param[in]  nupds	the count of SWIM updates
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/include/cart/swim.h:197]: Agree, I miss this. Let's fix this in separate patch.

File: src/include/cart/swim.h [hunk 14]
```diff
@@ -163,15 +178,43 @@ void swim_self_set(struct swim_context *ctx, swim_id_t self_id);
...
+ * @param[in]  id	IDs of selected target for message
+ * @param[in]  to	IDs of message to send to
+ * @returns		0 on success, negative error ID otherwise
+ */
+int swim_updates_send(struct swim_context *ctx, swim_id_t id, swim_id_t to);
+
+int swim_ipings_suspend(struct swim_context *ctx, swim_id_t from_id,
+			swim_id_t to_id, void *args);
+
+int swim_ipings_reply(struct swim_context *ctx, swim_id_t from_id, int ret_rc);
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/include/cart/swim.h:216]: The same here.

**Pull request merged**